thirdweb/modules/base.py

```diff
@@ -108,7 +108,7 @@ def upload_metadata(self, data: Union[Dict, str]) -> str:
         Uploads the metadata to IPFS and returns the uri.
         """
         storage = self.get_storage()
-        if isinstance(data, str) and data.startswith("ipfs://"):
+        if isinstance(data, str):
```

Comment:
Remove this entire if statement. In the typescript repo, the upload function simply uploads everything passed to it.

https://github.com/nftlabs/nftlabs-sdk-ts/blob/main/src/common/ipfs.ts#L38

Reply:
Also we should remove the statement here too https://github.com/nftlabs/nftlabs-sdk-python/blob/master/thirdweb/modules/base.py#L111

The entire problem with this *check if its a string* is because the SDK **does not enforce that metadata being provided is intended to be uploaded**.

I've brought up that all this implicit behavior in the SDKs not only causes the logic to be messy, but also makes it super confusing for the caller. This is a perfect example of that https://github.com/nftlabs/nftlabs-sdk-ts/blob/main/src/common/ipfs.ts#L66.

So as of right now, the callers should expect that **everything they pass to the sdk will be uploaded to storage**. Whereas before, we would say "if you pass a URI, we won't upload the data." But we're not actually checking if anything is a URI anywhere in any of the SDKs.

cc: @furqanrydhan @jakeloo

Reply:
Also @ayshptk we can change the sig for this function https://github.com/nftlabs/nftlabs-sdk-python/blob/master/thirdweb/modules/base.py#L106

```py
    def upload_metadata(self, data: Dict) -> str:
```

We won't ever be passing a string to upload.

Reply:
> The entire problem with this check if its a string is because the SDK does not enforce that metadata being provided is intended to be uploaded.

Is it a problem or it's a solution implemented poorly? We want to support the use case of letting people upload whatever they want. We shouldn't be opinionated on data storage layer and let people bring whatever they want, if they want too. If they don't, we provide a good default. 
That said, if I pass in "sia://" "arweave://" the SDK shouldn't just fail. Ultimately, all i'm trying to do is mint nft

Reply:
> So as of right now, the callers should expect that everything they pass to the sdk will be uploaded to storage. Whereas before, we would say "if you pass a URI, we won't upload the data." But we're not actually checking if anything is a URI anywhere in any of the SDKs.

I think this is either separate function or docs come in. I don't think we should do any validation but rather hands it off to people know what they're doing -- to accommodate for advanced users.

Reply:
>Remove this entire if statement. In the typescript repo, the upload function simply uploads everything passed to it.

This sounds problematic and could potentially caused problem in the downstream code path.

Reply:
> Is it a problem or it's a solution implemented poorly?

Its a case of a poorly designed SDK interface from the beginning. All of the type unions on the typescript sdk make it impossible to know whats being passed without doing type-checks, and for the caller, its even more confusing because they get different behavior based on whats passed to the function.

> That said, if I pass in "sia://" "arweave://" the SDK shouldn't just fail. Ultimately, all i'm trying to do is mint nft

Again, we need to take an explicit stance on whether or not a function is *intended to upload to storage*. Right now, our functions are designed such that someone *could pass either a url to a file OR some data that they want to upload*. This is the poor design I'm referencing, and the reason for these URL checks, because if someone passes a URL, we shouldn't be uploading that to IPFS!

> I think this is either separate function or docs come in. I don't think we should do any validation but rather hands it off to people know what they're doing -- to accommodate for advanced users.

Having this be a separate function is the clearest approach IMO and what I was advocating for from the beginning of the SDKs writing.

> This sounds problematic and could potentially caused problem in the downstream code path. Are we sure we've thought through this change?

This will cause 0 issues in the downstream code. I would not suggest we make the change if I didn't know what I was talking about. All calls to this function are internal and they've all been checked to ensure that we're passing a `Dict`.

Reply:
> Its a case of a poorly designed SDK interface from the beginning. All of the type unions on the typescript sdk make it impossible to know whats being passed without doing type-checks, and for the caller, its even more confusing because they get different behavior based on whats passed to the function.

The reason why I mentioned this is that we modeled all the other SDKs around the typescript SDK, including the python SDK. But as it stands, the typescript SDK is unclear on the following:

- What parameters can be either an object that *should be uploaded to storage* and what parameters allow a URL to be passed which will skip file upload
- What properties are expected for metadata and what properties are optional

Reply:
> This is the poor design I'm referencing, and the reason for these URL checks, because if someone passes a URL, we shouldn't be uploading that to IPFS.
> This will cause 0 issues in the downstream code. I would not suggest we make the change if I didn't know what I was talking about. All calls to this function are internal and they've all been checked to ensure that we're passing a `Dict`.
By removing this if check, the code would tries to upload string to `upload.nftlabs.co`, which is not ideal. right? https://github.com/nftlabs/nftlabs-sdk-ts/blob/ee7b92064ef7f7ff23dd002654ab958f17ed9f09/src/common/ipfs.ts#L66-L73

Reply:
> Having this be a separate function is the clearest approach IMO and what I was advocating for from the beginning of the SDKs writing.

Yup, I think that's fair. Worth thinking about how the interfaces would look like.
1. We want to support nft metadata dictionary, which we will upload and pin to ipfs for them.
2. Support their own url. we don't care if it's on ipfs, arweave, s3, gcloud, etc. We don't upload to IPFS.